### PR TITLE
Add support for ratelimiting via annotations

### PR DESCRIPTION
**Pull request by LukasReschke**
This allows adding rate limiting via annotations to controllers, as one example:

```
@UserRateThrottle(limit=5, period=100)
@AnonRateThrottle(limit=1, period=100)
```

Would mean that logged-in users can access the page 5 times within 100 seconds, and anonymous users 1 time within 100 seconds. If only an AnonRateThrottle is specified that one will also be applied to logged-in users.

At the moment this only works if a memory cache has been configured such as Redis for performance reasons. It is however implemented in an interface and if someone fancies writing a slow SQL database follow-up PRs are welcome :)

Still need to fix some small tests. :)

### CODE REVIEW

File: apps/federatedfilesharing/lib/Controller/MountPublicLinkController.php [hunk 1]
```diff
@@ -120,7 +120,7 @@ public function __construct($appName,
 	 *
 	 * @NoCSRFRequired
 	 * @PublicPage
-	 * @BruteForceProtection publicLink2FederatedShare
+	 * @BruteForceProtection(action=publicLink2FederatedShare)
```
Comment (LukasReschke) [inline apps/federatedfilesharing/lib/Controller/MountPublicLinkController.php]: This has been introduced by @schiessle into master with https://github.com/nextcloud/server/pull/3120, it is however not flexible enough which is why I extended it to use key-value pairs and adjusted all existing usages.

File: lib/private/AppFramework/DependencyInjection/DIContainer.php [hunk 2]
```diff
@@ -53,11 +53,13 @@
 use OCP\Files\Folder;
 use OCP\Files\IAppData;
 use OCP\IL10N;
+use OCP\IMemcache;
 use OCP\IRequest;
 use OCP\IServerContainer;
 use OCP\IUserSession;
 use OCP\RichObjectStrings\IValidator;
 use OCP\Util;
+use SearchDAV\XML\Limit;
```
Comment (LukasReschke) [inline lib/private/AppFramework/DependencyInjection/DIContainer.php]: IDE fail :)

File: lib/private/AppFramework/DependencyInjection/DIContainer.php [hunk 3]
```diff
@@ -162,14 +164,29 @@ public function __construct($appName, $urlParams = array(), ServerContainer $ser
 			return $c->query(Validator::class);
 		});
 
+		$this->registerService(OC\Security\RateLimiting\Limiter::class, function($c) {
```
Comment (rullzer) [inline lib/private/AppFramework/DependencyInjection/DIContainer.php]: Any particular reason not to register those in the servercontainer?

File: lib/private/AppFramework/DependencyInjection/DIContainer.php [hunk 4]
```diff
@@ -162,14 +164,29 @@ public function __construct($appName, $urlParams = array(), ServerContainer $ser
...
+			return new OC\Security\RateLimiting\Limiter(
+				$this->getServer()->getUserSession(),
+				$this->getServer()->getRequest(),
+				new OC\AppFramework\Utility\TimeFactory(),
+				$c->query(OC\Security\RateLimiting\Backend\IBackend::class)
+			);
+		});
+
+		$this->registerService(OC\Security\RateLimiting\Backend\IBackend::class, function($c) {
```
Comment (rullzer) [inline lib/private/AppFramework/DependencyInjection/DIContainer.php]: Any particular reason not to register those in the servercontainer?

File: lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php [hunk 5]
```diff
@@ -90,6 +91,8 @@ class SecurityMiddleware extends Middleware {
 	private $cspNonceManager;
 	/** @var Throttler */
 	private $throttler;
+	/** @var Limiter */
+	private $limiter;
```
Comment (rullzer) [inline lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php]: Would it make sense to have a separate middleware for this? Just for code complexity/manageablity etc? Or it is crucial to have it in the same?

File: lib/private/Security/Normalizer/IpAddress.php [hunk 6]
```diff
@@ -0,0 +1,106 @@
+<?php
```
Comment (rullzer) [inline lib/private/Security/Normalizer/IpAddress.php]: :rocket: nice! Much better this way!

File: lib/private/Security/RateLimiting/Backend/MemoryCache.php [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+		return $count;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function registerAttempt($methodIdentifier, $userIdentifier, $timestamp) {
+		$identifier = $this->hash($methodIdentifier, $userIdentifier);
+		$existingAttempts = $this->getExistingAttempts($identifier);
+		$existingAttempts[] = (string)$timestamp;
```
Comment (rullzer) [inline lib/private/Security/RateLimiting/Backend/MemoryCache.php]: Shouldn't you filter here? Else this juts keeps growing and growing?

**Review by rullzer — DISMISSED** [review]
haven't tested yet. But looks awesome already! :rocket:

File: lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php (hunk 5 above)
Comment (LukasReschke) [inline lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php]: Good point. Will add it's own middleware.

File: .drone.yml [hunk 8]
```diff
@@ -277,6 +277,17 @@ pipeline:
     when:
       matrix:
         TESTS: integration-maintenance-mode
+  integration-ratelimiting:
+  image: nextcloudci/integration-php7.0:integration-php7.0-3
```
Comment (MorrisJobke) [inline .drone.yml]: Indentation off 😉

File: apps/federatedfilesharing/lib/Controller/MountPublicLinkController.php (hunk 1 above)
Comment (rullzer) [inline apps/federatedfilesharing/lib/Controller/MountPublicLinkController.php]: looks good to me. Somewhat related. We should document those annotations properly ;)

**Review by rullzer — APPROVED** [review]
awesome stuff!

**Comment by LukasReschke** [comment]
@rullzer Yes. Will document once this is merged or we agreed on the annotations :-)

Will also make sure the BruteforceProtection is properly documented.

**Comment by LukasReschke** [comment]
Rebased to make Drone reprocess this. :)

**Comment by LukasReschke** [comment]
@BernhardPosselt Any thoughts about this? :)

File: apps/federatedfilesharing/lib/Controller/MountPublicLinkController.php (hunk 1 above)
Comment (LukasReschke) [inline apps/federatedfilesharing/lib/Controller/MountPublicLinkController.php]: Documentation PR at https://github.com/nextcloud/documentation/pull/411

**Review by ChristophWurst — APPROVED** [review]
Code looks very good! 🚀

**Pull request merged**